otlp/common.go

```diff
@@ -402,33 +402,56 @@ func getMarshallableValue(value *common.AnyValue) interface{} {
 	return nil
 }
 
-// This function returns a value that can be handled by Honeycomb -- it must be one of:
-// string, int, bool, float. All other values are converted to strings containing JSON.
-func getValue(value *common.AnyValue) (result interface{}, truncatedBytes int) {
+func flatten(result map[string]interface{}, key string, value *common.AnyValue, depth int) (truncatedBytes int) {
 	switch value.Value.(type) {
 	case *common.AnyValue_StringValue:
-		return value.GetStringValue(), 0
+		result[key] = value.GetStringValue()
 	case *common.AnyValue_BoolValue:
-		return value.GetBoolValue(), 0
+		result[key] = value.GetBoolValue()
 	case *common.AnyValue_DoubleValue:
-		return value.GetDoubleValue(), 0
+		result[key] = value.GetDoubleValue()
 	case *common.AnyValue_IntValue:
-		return value.GetIntValue(), 0
-	// These types are all be marshalled to a string after conversion to Honeycomb-safe values.
-	// We use our limitedWriter to ensure that the string can't be bigger than the allowable,
-	// and it also minimizes allocations.
-	// Note that an Encoder emits JSON with a trailing newline because it's intended for use
-	// in streaming. This is correct but sometimes surprising and the tests need to expect it.
-	case *common.AnyValue_ArrayValue, *common.AnyValue_KvlistValue, *common.AnyValue_BytesValue:
-		arr := getMarshallableValue(value)
-		w := newLimitedWriter(fieldSizeMax)
-		enc := json.NewEncoder(w)
-		err := enc.Encode(arr)
-		if err == nil {
-			return w.String(), w.truncatedBytes
+		result[key] = value.GetIntValue()
+	case *common.AnyValue_BytesValue:
+		truncated := addToMapAsJson(result, key, value)
+		truncatedBytes += truncated
+	case *common.AnyValue_ArrayValue:
+		for i, entry := range value.GetArrayValue().Values {
+			k := fmt.Sprintf("%s.%d", key, i)
+			if depth < maxDepth {
+				flatten(result, k, entry, depth+1)
+			} else {
+				truncatedBytes += addToMapAsJson(result, key, entry)
+			}
+		}
+	case *common.AnyValue_KvlistValue:
+		for _, entry := range value.GetKvlistValue().Values {
+			k := fmt.Sprintf("%s.%s", key, entry.Key)
```

Comment:
In general, using Sprintf to concatenate things that are known to be strings is WAAAY more expensive than doing it with `+`, or for more complicated stuff, strings.Builder. It'd be worth changing this to just `+`.

Here's a benchmark:
```
BenchmarkSprintf-10          	  817777	      1474 ns/op	     108 B/op	       3 allocs/op
BenchmarkStringCat-10        	41451492	        29.89 ns/op	       0 B/op	       0 allocs/op
BenchmarkStringBuilder-10    	 2964910	       400.9 ns/op	      32 B/op	       2 allocs/op
```

```
var s1 = "prefix"
var s2 = "suffix"

func BenchmarkSprintf(b *testing.B) {
	for i := 0; i < b.N; i++ {
		_ = fmt.Sprintf("%s.%s", s1, s2)
	}
}

func BenchmarkStringCat(b *testing.B) {
	for i := 0; i < b.N; i++ {
		_ = s1 + "." + s2
	}
}

func BenchmarkStringBuilder(b *testing.B) {
	for i := 0; i < b.N; i++ {
		var sb strings.Builder
		sb.WriteString(s1)
		sb.WriteRune('.')
		sb.WriteString(s2)
		_ = sb.String()
	}
}
```

Reply:
I know I knew this, but always find it hard to not try to avoid using `+` for string concatenation :|

Reply:
Fixed in https://github.com/honeycombio/husky/pull/240/commits/70e188692fe32631aafc252e4f59bc86ac050fd4